### DISCUSSION TITLE
How do I get MAC and IP addresses associated with a client socket returned from socket.accept()?

### DISCUSSION THREAD

**Discussion by jimkoke** [micropython/micropython#13353] [Libraries & Drivers]
I've got an application running on Pi-Pico that listens for client connections and performs various tasks. In order to perform some of these, I need the client host's MAC and IP addresses. I know I can ask the the client to provide these as part of the request, however, I believe I should be able to extract these from the client socket provided by the accept() function, which simplifies the API and streamlines the requests.

I have searched extensively (but apparently not enough) and found several solutions, however, those work for "Python" but do not work for micropython.

The code uses "socket.socket()" to get the service's socket, the listens on that socket, finally, accept() to get the client socket. The code is not so concise, I would have to cobble it together if need by, but it seems fairly standard, taking the code from Internet examples.

[UPDATE}
After reading sophiedegran's question, and smacking myself on the forehead, I realized that I have the requestor's IP address already. I did not realize it because I blindly copy-n-pasted code without reading it thoroughly:

    cl, addr = listenerSocket.accept()

I'm adding this line to get the requestor's IP and port

    clientIp, clientPort = addr

I am now left with finding the requestor's MAC address. All of the answers I have found use ARP, which might work, if these IOT processors would respond. It just seems there would be some access to the MAC layer of the incoming packet.

**Comment by sophiedegran** [comment]
Hi
You have to do first a socket.accept() to have any informations..
After you do the restriction..
I don't know how your socket is open on your server so it's difficult to give you more info. for ip..

**Reply by jimkoke** [reply]
After smacking my forehead, I updated my original post, thanks for the wakeup, still need to find the MAC address.

**Comment by Josverl** [comment]
Are you assuming that the client is on the same physical network? 
Assuming a routed network,the originator MAC address could be a HAM callsign on packet radio, or a token ring address, or the IMEI on a phones cellular interface.

In these cases the most revenant MAC address is that of the routers interface that you are sending your reply to.


[Some relevant words](https://networkengineering.stackexchange.com/questions/23351/how-are-mac-ip-addresses-used-in-routing)

**Reply by sophiedegran** [reply]
Yes if you want to 'filtering' your socket server and client access the best thing to do is to encrypt your message with a secret_key
See pyjwt (jwt) in micropython-lib/python-ecosys example .. 
If the sender don't have the good key you simply close the socket connexion ...
